pydantic/main.py

```diff
@@ -246,7 +244,15 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__repr__': Representation.__repr__,
             **{n: v for n, v in namespace.items() if n not in fields},
         }
-        return super().__new__(mcs, name, bases, new_namespace, **kwargs)
+        cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
+        if _custom_root_type and cls.__fields__[ROOT_KEY].shape == SHAPE_MAPPING:
```

Comment:
The check for the `SHAPE_MAPPING` could go into the `parse_obj` method definition, but then you'd pay the overhead on every call. 

@samuelcolvin I'm guessing you might find that preferable since it's easier to trace the logic, but I figured I'd propose this approach in case you prefer the performance benefits.

If you like this approach, I think we could actually move all of the `_custom_root_type` checks here instead of in the method definition to reduce the number of checks used when working with non-custom-root-type models, but it quickly makes it harder to follow the logic of the method if there is a custom root type.

Maybe it would be okay if we factored this logic out of the metaclass `__new__` and into a separate `_setup_custom_root` function, and just added a reference to that function in the `parse_obj` docstring?

Reply:
Interesting idea, but I think too confusing to avoid one `if <boolean>`.

I think:
```py
    @classmethod
    def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
        if cls.__custom_root_type__:
            obj = {ROOT_KEY: obj}
        elif not isinstance(obj, dict):
            try:
                obj = dict(obj)
            except (TypeError, ValueError) as e:
                exc = TypeError(f'{cls.__name__} expected dict not {type(obj).__name__}')
                raise ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls) from e
        return cls(**obj)
```

Would be better.

---

But this (and your solution) lead to a backwards incompatibility:

If you call `parse_obj({'__root__': whatever})` validation would now fail since you'd really be calling 

```py
Model(**{'__root__': {'__root__': whatever}})`
```

I think this confusion is why we introduced the check on root type in the first place. We should have changed this pre-v1, but too late.

Solutions:
* wait for version 2 to make this change :-(
* change the above function definition to 

```py
    @classmethod
    def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
        if cls.__custom_root_type__ and not (isinstance(obj, dict) and obj.keys() == {'__root__'}):
        ...
```
That wouldn't be any slower for the standard case, but would involve some more logic if `__custom_root_type__ = True`. I think that would be best for now.

We might want to leave a note saying we'll remove the `and not (isinstance(obj, dict) and obj.keys() == {'__root__'})` in future.

Thoughts?

Reply:
I actually think this is a really clean way to handle this; I didn't think about the fact that the check is always being performed now anyway.

I think I'd be in favor of modifying this behavior in v2, but we can worry about that when the time comes.